src/ppx/xr_component.cpp

```diff
@@ -48,6 +48,43 @@ static XrBool32 XrDebugUtilsMessengerCallback(XrDebugUtilsMessageSeverityFlagsEX
     return XR_FALSE; /* OpenXR spec suggests always returning false. */
 }
 
+template <int N>
+void CharArrayCopy(char (&dst)[N], const char* src)
+{
+    strncpy(dst, src, N);
+    // Force the last byte to be '\0'
+    dst[N - 1] = '\0';
+}
+
+float Sign(float x)
+{
+    return (x < 0 ? -1 : (x > 0 ? 1 : 0));
+}
+
+XrVector3f ForwardVector(const XrPosef& controller)
+{
+    XrQuaternionf q = controller.orientation;
+    return XrVector3f{
+        -2.0f * (q.x * q.z + q.w * q.y),
+        -2.0f * (q.y * q.z - q.w * q.x),
+        2.0f * (q.x * q.x + q.y * q.y) - 1.0f,
+    };
+}
+
+std::optional<XrVector2f> ProjectCursor(XrPosef controller, float z)
+{
+    XrVector3f pos = controller.position;
+    XrVector3f fwd = ForwardVector(controller);
+
+    float zDistance = z - pos.z;
+    // Controller is not pointing forward (or very sideway)
+    if (Sign(zDistance) * fwd.z < 0.001) {
+        return std::nullopt;
+    }
+    float scalar = zDistance / fwd.z;
```

Comment:
is it possible that fwd.z is 0?

Reply:
`Sign(zDistance) * fwd.z == 0` if `fwd.z == 0`